[release/8.0-staging] Fix #91958: use mach_timebase_info to determine process time coefficient on macOS

jeffhandley opened the pull request:
Backport of #92185 to release/8.0-staging

## Customer Impact

- [x] Customer reported
- [ ] Found internally

This has been reported by 2 customers. The first report was in September 2023 by @ForNeVeR in #91958, and they also provided the fix and tests. The issue was reported again by @huntharo in February with #98121. In that issue, @huntharo requested that this fix be backported to 8.0. As @huntharo stated:

> On Mac OS X, with Apple Silicon, the `System.Diagnostics.Process` class returns incorrect (substantially underreported - ms when it should be seconds) CPU usage times for the process (and probably for the thread-level stats too).
>
> I discovered this while writing a function to tune the max # of worker threads in the ThreadPool as it reduces CPU usage from 700% to 120% while increasing throughput for my project (https://github.com/pwrdrvr/lambda-dispatch/issues/109).  I was always computing that 0 threads would be needed because CPU usage was only going up by 20 ms per 5 seconds despite using 700% CPU.

## Regression

- [ ] Yes
- [x] No

While this isn't a product regression, it's behavior that became incorrect with Apple Silicon.

## Testing

@ForNeVeR included a new test with their fix that is part of this backport. The test performs a native call and compares the API's results against the native results. As stated in the original PR:

> I have compared the results of the process timing functions to the results of ps on my M2 MacBook device, and they seem to work properly after the changes.
>
> The new tests were properly failing before the change (since we were returning values 42 times lower than the native results), and are, of course, green after the changes.

## Risk

**Low.** This is a targeted fix that only affects `PrivilegedProcessorTime`, `TotalProcessorTime`, and `UserProcessorTime` on macOS. While the fix had been in main since September, a race condition bug in the original fix was found during the review of this backport. That bug was [fixed in main on March 25](https://github.com/dotnet/runtime/pull/100260) and cherry-picked to this backport.

The automatic backport of this fix failed due to a merge conflict in [src/libraries/Common/src/Interop/OSX/Interop.Libraries.cs](https://github.com/dotnet/runtime/blob/release/8.0-staging/src/libraries/Common/src/Interop/OSX/Interop.Libraries.cs). That conflict arose because the preceding line to the change here was removed in [20a3350](https://github.com/dotnet/runtime/commit/20a3350eeddf6a4d24675976586683193c1c1d1d). Otherwise, the rest of the merge was clean.

ForNeVeR committed:
Fix #91958: use mach_timebase_info to determine process time coefficient on macOS

ForNeVeR committed:
(#91958) Process.OSX: fallback for non-zero exit code from mach_timebase_info

ForNeVeR committed:
(#91958) Process: move OSX-specific tests to a separate file

ForNeVeR committed:
(#91958) Process.OSX: calculate times lazily

ForNeVeR committed:
(#91958) Process: include string resources properly in tests

ForNeVeR committed:
(#91958) Process: more overflow-safe logic

ForNeVeR committed:
(#91958) Minimize diff

adamsitnik reviewed (approved):
LGTM, thank you for backporting the fix @jeffhandley !

jkotas commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.OSX.cs:
```diff
@@ -109,5 +110,35 @@ private static Interop.libproc.rusage_info_v3 GetCurrentProcessRUsage()
+            if (denom == default)
+            {
+                Interop.libSystem.mach_timebase_info_data_t timeBase = GetTimeBase();
+                s_timeBase_denom = denom = timeBase.denom;
+                s_timeBase_numer = timeBase.numer;
```
```suggestion
                s_timeBase_numer = timeBase.numer;
                s_timeBase_denom = denom = timeBase.denom;
```
This has a race condition. Consider what happens if the thread is rescheduled right after `s_timeBase_denom` is assigned and before `s_timeBase_numer` is assigned; and some other thread executes this method. The `denom == default` condition is going to be false on the other thread and it will end up uninitialized `s_timeBase_numer`.

ForNeVeR replied:
Yep, you are right. No idea how we missed that during the initial review :(

What is the process to apply the fix for this race to both `master` and now to .NET 8 as well?

jkotas replied:
Submit a PR with the fix to against the main branch. We can then cherry pick it into this one.

jkotas replied:
Cherry-picked the fix from main

ForNeVeR committed:
Process.OSX: fix thread safety (#100260)

Without this fix, it was possible for another thread to see an incorrect
(zero) value of s_timeBase_numer because of a race condition.

Jozkee commented:
@jeffhandley should this be marked as servicing consider now that https://github.com/dotnet/runtime/pull/100122#discussion_r1536727982 got resolved? If so, do we need to email tactics?

carlossanlop commented:
@Jozkee I didn't find a Tactics email, so assuming this is ready (pending @jeffhandley 's or @adamsitnik 's confirmation), then yes, please send an email to Tactics and add the `servicing-consider` label.

jeffhandley commented:
This was approved in email and needs an approval for merge. @jkotas -- you can ignore the re-review request as this can now be merged.
